### PR TITLE
Use CMake for building & running CBMC proofs

### PR DESCRIPTION
**Pull request by karkhaz**
This PR introduces a CMake workflow for building and running CBMC proofs. It includes the CMake infrastructure shared by all proofs, as well as converting the current HTTP proofs to use CMake rather than the previous python-based build flow.

This PR is ready for discussion about the general approach and where the various files should live. It works for building proofs locally, but the proofs do not currently run in CI. I'll fix that before requesting final approval.

@tgsong and @markrtuttle: are you happy with the following?
- `cbmc.cmake` lives under `tools/cmake/toolchains`
- `cbmc_recipes.cmake` lives under `tools/cmake` (I guess it could also live anywhere under `tools/cbmc`)
- The CMakefiles for the individual proofs (as well as the proofs themselves) all live somewhere beneath `tools/cbmc/proofs`. They could, instead, live right next to the code that they are proving. For example, the HTTP proofs could live under `libraries/c_sdk/standard/https/cbmc`, which would mean that the HTTP source code, tests, and proofs all live under the `https` directory together.

**Note to reviewers**: I suggest reading in the following order:
- tools/cbmc/README.md
- The CMakeLists.txt for a single proof, e.g. `IotHttpsClient_AddHeader/CMakeLists.txt`
- `tools/cmake/cbmc_recipies.cmake`, which gets included by every single proof file and is responsible for generating the build & run for the proof
- `tools/cmake/toolchains/cbmc.cmake`, which is processed once, before any proof CMakeLists.txt gets read, and which sets up the CBMC toolchain and defines variables that proofs can subsequently use
- @markrtuttle please also review all other`CMakeLists.txt`s to ensure that they correspond to the deleted `Makefile.json`s for each proof.

Things to complete before I take this out of draft mode:

- [ ] Resolve all TODOs
- [ ] Make the CI also build and run the CMake proofs

### CODE REVIEW

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,45 @@
+set(cbmc_proof_name IotHttpsClient_ReadHeader)
+list(APPEND cbmc_test_labels ${cbmc_proof_name})
+
+set(IotHttpsClient_ReadHeader_MAX_ACCEPTED_SIZE 20 CACHE STRING "TODO what is this?")
```
Comment (karkhaz) [inline tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt]: @markrtuttle please could you let me know what a better documentation string for this variable would be?

File: tools/cmake/cbmc_recipes.cmake [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+list(LENGTH cbmc_unwindset cbmc_unwindset_length)
+if(${cbmc_unwindset_length} GREATER 0)
+    list(JOIN cbmc_unwindset "," _cbmc_unwindset)
+    list(APPEND cbmc_flags --unwindset ${_cbmc_unwindset})
+endif()
+
+list(JOIN cbmc_flags " " _cbmc_flags)
+
+# Actually check the proof
+# TODO this obviously won't work on windows.
```
Comment (gkwicker) [inline tools/cmake/cbmc_recipes.cmake]: Does any of this infrastructure work on Windows?  If not, I'd delete the comments at lines 254, 269, 283, and 310 and add a check for the cmake `UNIX` variable at the top of this file; if it's not defined then you can notify the user that it's not supported on other OSes.

```
if( NOT UNIX )
  message( FATAL_ERROR "Building and running CBMC proofs is only supported on UNIX-type systems" )
endif()

```

Comment (karkhaz) [reply tools/cmake/cbmc_recipes.cmake]: Thanks @gkwicker! We're definitely planning to make it work on Windows, we'd rather not merge it before that point. I'll take this PR out of draft mode once I fix this.

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt (hunk 1 above)
Comment (markrtuttle) [reply tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt]: This variable appears to be unneeded in the proof.

File: tools/cmake/cbmc_recipes.cmake [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+if(${cnse_length} AND ${cbmc_nondet_static})
+    message(FATAL_ERROR "\
+The CMakeLists.txt for the CBMC proof in directory
+${CMAKE_CURRENT_SOURCE_DIR}
+sets both `cbmc_nondet_static` to true, and `cbmc_nondet_static_exclude` to a
+nonempty list. You should only do one or the other. Value of
+cbmc_nondet_static_exclude: [${cbmc_nondet_static_exclude}].")
+endif()
+
+set(cbmc_proof_names "${cbmc_proof_name};${cbmc_proof_names}" CACHE INTERNAL "")
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: Should this be list append?

File: tools/cmake/cbmc_recipes.cmake [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+    else()
+      set(cmd_list
+          cmake -E copy
+          ${cbmc_proof_name}_0010_${binary}.goto
+          ${cbmc_proof_name}_0020_${binary}_functions_removed.goto
+      )
+    endif()
+
+    add_custom_command(
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: I recommend moving this custom command definition inside the if-then-else.  Then the command can be easily stated (just reference the interpolated list), and the comment can be set to something sensible.

File: tools/cmake/cbmc_recipes.cmake [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    OUTPUT
+        ${cbmc_proof_name}_0050_unused_functions_dropped.goto
+    COMMAND
+        goto-instrument --drop-unused-functions --verbosity ${CBMC_VERBOSITY}
+        ${cbmc_proof_name}_0040_library_added.goto
+        ${cbmc_proof_name}_0050_unused_functions_dropped.goto
+)
+
+list(LENGTH cbmc_nondet_static_exclude cnse_length)
+if(${cbmc_nondet_static} OR ${cnse_length})
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: Comment that you know exactly one is true given the sanity checking at top level of this file.

File: tools/cmake/cbmc_recipes.cmake [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        COMMAND
+            goto-instrument ${nondet_static_flags}
+            --verbosity ${CBMC_VERBOSITY}
+            ${cbmc_proof_name}_0050_unused_functions_dropped.goto
+            ${cbmc_proof_name}_0060_nondet_static.goto
+    )
+else()
+    add_custom_command(
+        COMMENT
+            "Not havoccing statics in ${cbmc_proof_name}"
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: This should be havocked

File: tools/cmake/cbmc_recipes.cmake [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+            ";--nondet-static-exclude;" nondet_static_flags)
+        list(PREPEND
+            nondet_static_flags
+            --nondet-static-exclude
+        )
+    endif()
+
+    add_custom_command(
+        COMMENT
+            "Havoccing all except ${cnse_length} statics in ${cbmc_proof_name}"
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: havocked

File: tools/cmake/cbmc_recipes.cmake [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    --unwind ${cbmc_unwind}
+)
+
+list(LENGTH cbmc_unwindset cbmc_unwindset_length)
+if(${cbmc_unwindset_length} GREATER 0)
+    list(JOIN cbmc_unwindset "," _cbmc_unwindset)
+    list(APPEND cbmc_flags --unwindset ${_cbmc_unwindset})
+endif()
+
+list(JOIN cbmc_flags " " _cbmc_flags)
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: Just pass as a list not as a string.  No need for join.

File: tools/cmake/cbmc_recipes.cmake [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+# Actually check the proof
+# TODO this obviously won't work on windows.
+add_test(
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: Assuming the rest of this is going to be replaced with your new implementation that is OS agnostic.

File: tools/cmake/cbmc_recipes.cmake [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+)
+
+# Accumulate the results of the proof runs above and generate a report
+set(_report_command
+    cbmc-viewer
+    --goto ${cbmc_proof_name}.goto
+    --srcdir ${CMAKE_SOURCE_DIR}
+    --blddir ${CMAKE_SOURCE_DIR}
+    --htmldir ${CMAKE_BINARY_DIR}/cbmc-reports/${cbmc_proof_name}
+    --srcexclude "\"(./doc|./tests|./vendors)\""
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: This list of directories to exclude should not be hardwired.

File: tools/cmake/cbmc_recipes.cmake [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+    --result cbmc.txt
+    --property property.xml
+    --block coverage.xml
+)
+list(JOIN _report_command " " report_command)
```
Comment (markrtuttle) [inline tools/cmake/cbmc_recipes.cmake]: Treat this as a list, no need to join.

File: tools/cmake/toolchains/cbmc.cmake [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+#  0 no output
+#  1 errors
+#  2 warnings
+#  4 results
+#  6 status information
+#  8 statistical information
+#  9 progress information
+# 10 debug info
+set(CBMC_VERBOSITY 4 CACHE STRING "Verbosity of CBMC and related tools")
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: Use the CBMC default of 6 for the default.

File: tools/cmake/toolchains/cbmc.cmake [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+# ______________________________________________________________________________
+#     Variables that are not intended to be directly changed by proof-writers
+# ``````````````````````````````````````````````````````````````````````````````
+
+# The list of all proof names (used to check that proof names are unique)
+set(cbmc_proof_names "" CACHE INTERNAL "")
+
+# The cbmc root
+set(cbmc_dir ${CMAKE_SOURCE_DIR}/tools/cbmc)
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: These dir names are FreeRTOS and MQTT specific, crypto uses different stuff.

File: tools/cmake/toolchains/cbmc.cmake [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+# ______________________________________________________________________________
+#     Strings that proof-writers can change for a particular proof in each
+#     proof's CMakeLists.txt
+# ``````````````````````````````````````````````````````````````````````````````
+
+# Argument to cbmc --unwind flag. See also cbmc_unwindset below.
+set(cbmc_unwind 1)
+
+# Setting this to true for a particular proof builds that proof with the
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: Use ON/OFF or TRUE/FALSE.

File: tools/cmake/toolchains/cbmc.cmake [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+set(cbmc_compile_options)
+set(cbmc_compile_definitions)
+set(cbmc_compile_includes)
+set(cbmc_flags)
+
+# ______________________________________________________________________________
+#     Use goto-cc (the CBMC compiler) and do other FreeRTOS-specific setup
+# ``````````````````````````````````````````````````````````````````````````````
+
+include("${CMAKE_CURRENT_LIST_DIR}/find_compiler.cmake")
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: From here on is FreeRTOS specific, can this be made more generic?

File: tools/cmake/toolchains/cbmc.cmake [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+set(CMAKE_SYSTEM_NAME Generic)
+
+afr_find_compiler(AFR_COMPILER_CC goto-cc)
+afr_find_compiler(AFR_COMPILER_CXX goto-cc)
+set(AFR_COMPILER_ASM "${AFR_COMPILER_CC}" CACHE INTERNAL "")
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: Can we skip the paragraph above?  Can't we just set the compilers below to goto-cc?

File: tools/cmake/toolchains/cbmc.cmake [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+# Specify the cross compiler.
+set(CMAKE_C_COMPILER ${AFR_COMPILER_CC} CACHE FILEPATH "C compiler")
+set(CMAKE_CXX_COMPILER ${AFR_COMPILER_CXX} CACHE FILEPATH "C++ compiler")
+set(CMAKE_ASM_COMPILER ${AFR_COMPILER_ASM} CACHE FILEPATH "ASM compiler")
+
+# Disable compiler checks.
```
Comment (markrtuttle) [inline tools/cmake/toolchains/cbmc.cmake]: Is this comment right?

File: tools/cmake/cbmc_recipes.cmake (hunk 3 above)
Comment (karkhaz) [reply tools/cmake/cbmc_recipes.cmake]: It seems that `list(APPEND` doesn't work on cache variables

File: tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt (hunk 1 above)
Comment (karkhaz) [reply tools/cbmc/proofs/HTTP/IotHttpsClient_ReadHeader/CMakeLists.txt]: done

File: tools/cmake/cbmc_recipes.cmake (hunk 9 above)
Comment (karkhaz) [reply tools/cmake/cbmc_recipes.cmake]: done

File: tools/cmake/toolchains/cbmc.cmake (hunk 13 above)
Comment (karkhaz) [reply tools/cmake/toolchains/cbmc.cmake]: As agreed, let's merge this and refactor it later once it's clear what other projects will need

File: tools/cmake/toolchains/cbmc.cmake (hunk 15 above)
Comment (karkhaz) [reply tools/cmake/toolchains/cbmc.cmake]: As agreed, let's merge this and refactor it later once it's clear what other projects will need

**Comment by karkhaz** [comment]
@gkwicker this is now ready for review. I've made it cross-platform by wrapping the CBMC invocations in CMake scripts, run with `cmake -P`, rather than using the POSIX shell.

File: tools/cbmc/README.md [hunk 18]
```diff
@@ -13,16 +13,61 @@ analysis tool
...
+### Prerequisites
 
 You will need Python 3. On Windows, you will need Visual Studio 2015 or later
 (in particular, you will need the Developer Command Prompt and NMake). On macOS
 and Linux, you will need Make.
 
 
-Installing CBMC
----------------
+### Installing CBMC
```
Comment (dcgaws) [inline tools/cbmc/README.md]: These steps aren't working for me on native Windows. I do have win_bison.exe and win_flex.exe in the local directory.

C:\GitHub\cbmc>cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DWITH_JBMC=OFF
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.17763.
-- The C compiler identification is MSVC 19.23.28107.0
-- The CXX compiler identification is MSVC 19.23.28107.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.14.2.windows.1")
-- Could NOT find BISON (missing: BISON_EXECUTABLE)
-- Could NOT find FLEX (missing: FLEX_EXECUTABLE)
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
CMake Error at src/CMakeLists.txt:26 (bison_target):
  Unknown CMake command "bison_target".
Call Stack (most recent call first):
  src/ansi-c/CMakeLists.txt:1 (generic_bison)
-- Configuring incomplete, errors occurred!
See also "C:/GitHub/cbmc/build/CMakeFiles/CMakeOutput.log".

Comment (dcgaws) [reply tools/cbmc/README.md]: Here's the command-line that works for me on Windows:

`cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DWITH_JBMC=OFF -DBISON_EXECUTABLE=c:\GitHub\cbmc\win_bison.exe -DFLEX_EXECUTABLE=c:\GitHub\cbmc\win_flex.exe`

Comment (dcgaws) [reply tools/cbmc/README.md]: More progress, per offline discussion, now using Python 3.8.1 on Windows:

C:\GitHub\afr\tools\cbmc\proofs>python prepare.py
patch C:\GitHub\afr\tools\cbmc\patches\0001-Remove-static-storage-class-from-entry-points.patch
patch C:\GitHub\afr\tools\cbmc\patches\0002-Change-FreeRTOS_IP_Private.h-union-to-struct.patch
patch C:\GitHub\afr\tools\cbmc\patches\0005-remove-static-from-prvCopyDataToQueue.patch
error: patch failed: freertos_kernel/queue.c:105
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0005-remove-static-from-prvCopyDataToQueue.patch
patch C:\GitHub\afr\tools\cbmc\patches\0005-Remove-volatile-qualifier-from-tasks-variables.patch
error: patch failed: freertos_kernel/tasks.c:331
error: freertos_kernel/tasks.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0005-Remove-volatile-qualifier-from-tasks-variables.patch
patch C:\GitHub\afr\tools\cbmc\patches\0006-Remove-static-from-prvNotifyQueueSetContainer.patch
error: patch failed: freertos_kernel/queue.c:207
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0006-Remove-static-from-prvNotifyQueueSetContainer.patch
patch C:\GitHub\afr\tools\cbmc\patches\0007-Remove-static-from-prvUnlockQueue.patch
error: patch failed: freertos_kernel/queue.c:175
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0007-Remove-static-from-prvUnlockQueue.patch
patch C:\GitHub\afr\tools\cbmc\patches\0008-Remove-static-attributes-from-functions-implementing.patch

**Pull request merged**